### PR TITLE
Introduce a GetChannels method to the base plugin API

### PR DESCRIPTION
**Pull request by nadalfederer**
Summary
Add GetChannels with opts for plugin API

Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12418

### CODE REVIEW

**Comment by nadalfederer** [comment]
I fixed errors which were here https://github.com/mattermost/mattermost-server/pull/12649

File: app/plugin_api.go [hunk 1]
```diff
@@ -330,6 +330,10 @@ func (api *PluginAPI) GetChannel(channelId string) (*model.Channel, *model.AppEr
 	return api.app.GetChannel(channelId)
 }
 
+func (api *PluginAPI) GetChannels(options *model.GetChannelsOptions) (*model.ChannelList, *model.AppError) {
```
Comment (hanzei) [inline app/plugin_api.go]: Given that `options` is a required parameter, it should be a struct, not a pointer.

Comment (hanzei) [inline app/plugin_api.go]: Is there a need for returning `*model.ChannelList`? Why not just `[]*model.Channel`?

File: model/channel.go [hunk 2]
```diff
@@ -188,6 +201,18 @@ func (o *Channel) IsValid() *AppError {
 	return nil
 }
 
+func (o *GetChannelsOptions) IsValid() *AppError {
+	if reflect.TypeOf(o.UserIds).String() != "[]string" {
```
Comment (hanzei) [inline model/channel.go]: Can you elaborate why this is needed? The struct definition already ensures that `o.UserIds` is a string slice.

File: model/channel.go [hunk 3]
```diff
@@ -188,6 +201,18 @@ func (o *Channel) IsValid() *AppError {
...
+		return NewAppError("GetChannelsOptions.IsValid", "model.channel.is_valid.user.ids.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	if reflect.TypeOf(o.ChannelTypes).String() != "[]string" {
+		return NewAppError("GetChannelsOptions.IsValid", "model.channel.is_valid.channel.types.app_error", nil, "", http.StatusBadRequest)
+	}
+
```
Comment (hanzei) [inline model/channel.go]: Should the content of `ChannelTypes` and `UserIds` also be valided here? E.g. if the elements in `ChannelTypes` have the correct values like `model.CHANNEL*OPEN`.

File: store/sqlstore/channel_store.go [hunk 4]
```diff
@@ -873,6 +873,49 @@ func (s SqlChannelStore) GetChannels(teamId string, userId string, includeDelete
...
+
+	if err := opt.IsValid(); err != nil {
+		return nil, err
+	}
+
+	query := s.getQueryBuilder().
+		Select("c.*").
+		From("Channels AS c")
+
+	if opt.Page < 0 {
```
Comment (hanzei) [inline store/sqlstore/channel_store.go]: Why is this checked here and not in `IsValid()`?

File: app/plugin_api_test.go [hunk 5]
```diff
@@ -998,6 +998,45 @@ func TestPluginAPIGetChannelsForTeamForUser(t *testing.T) {
...
+		package main
+		import (
+			"github.com/mattermost/mattermost-server/v5/plugin"
+			"github.com/mattermost/mattermost-server/v5/model"
+		)
+		type MyPlugin struct {
+			plugin.MattermostPlugin
+		}
+		func (p *MyPlugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*model.Post, string) {
+			_, err := p.API.GetChannels(&model.GetChannelsOptions{
```
Comment (iomodo) [inline app/plugin_api_test.go]: Can we check what actually is returned by `GetChannels` method?

File: model/channel.go (hunk 2 above)
Comment (iomodo) [inline model/channel.go]: `IsValid` is a good idea, Can `o.UserIds` or `o.ChannelTypes` be anything other than `[]string` ?
I'd add checking `page` and `perPage` here.

File: plugin/api.go [hunk 6]
```diff
@@ -348,6 +348,12 @@ type API interface {
 	// Minimum server version: 5.2
 	GetChannel(channelId string) (*model.Channel, *model.AppError)
 
+	// GetChannels gets channels.
+	//
+	// @tag Channel
+	// Minimum server version: 5.21
```
Comment (iomodo) [inline plugin/api.go]: why 5.21?

File: store/sqlstore/channel_store.go [hunk 7]
```diff
@@ -873,6 +873,49 @@ func (s SqlChannelStore) GetChannels(teamId string, userId string, includeDelete
...
+		return nil, model.NewAppError("SqlChannelStore.GetChannelsWithOptions", "store.sql_channel.get_channels_with_options.get.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	if opt.PerPage < 0 {
+		return nil, model.NewAppError("SqlChannelStore.GetChannelsWithOptions", "store.sql_channel.get_channels_with_options.per_page_arg.app_error", nil, "", http.StatusBadRequest)
+	}
```
Comment (iomodo) [inline store/sqlstore/channel_store.go]: This can be in `IsValid`

File: store/sqlstore/channel_store.go [hunk 8]
```diff
@@ -873,6 +873,49 @@ func (s SqlChannelStore) GetChannels(teamId string, userId string, includeDelete
...
+
+	if len(opt.ChannelTypes) > 0 {
+		query = query.Where(fmt.Sprintf("c.Type IN ('%s')", strings.Join(opt.ChannelTypes, "', '")))
+	}
+
+	if len(opt.UserIds) > 0 {
+		query = query.Where(fmt.Sprintf("c.Id IN (SELECT DISTINCT(ChannelId) FROM ChannelMembers where UserId IN ('%s'))", strings.Join(opt.UserIds, "', '")))
+	}
+	if opt.Page >= 0 && opt.PerPage >= 0 {
```
Comment (iomodo) [inline store/sqlstore/channel_store.go]: No need of the check

**Comment by hanzei** [comment]
@nadalfederer Head up that your PR conflicts with https://github.com/mattermost/mattermost-server/pull/13302. That changed the way tests for plugins where written. It might take some time to resolve the conflict. Let me know, if you have any questions.

**Comment by jasonblais** [comment]
Hey @nadalfederer! Let us know if you have any questions about the feedback above?

**Comment by nadalfederer** [comment]
Hi, I'm sorry for the delay. I'll fix current issues soon

**Comment by nadalfederer** [comment]
This is still in progress.

**Comment by nadalfederer** [comment]
Hi, I cannot figure out/give up on this task. Can you unassign me from this task?

**Comment by hanzei** [comment]
Sure @nadalfederer. Just so I understand what the reason is: Did you needed more help with the resolving the review comments? Was the ticket unclear?

**Comment by nadalfederer** [comment]
I'm lacking understanding the go and codebase changed, I thought that I understand how to do it, but time went by and I forgot how it worked. I think that's it

**Comment by hanzei** [comment]
Thanks for sharing you thoughts @nadalfederer

**Pull request closed without merging**